File: insertion_sort/InsertionSort.py
def InsertionSort(arr: list):
    for i in range(1, len(arr)):
        j = i - 1
        temp = arr[i]
        while arr[j] > temp and j >= 0:
            arr[j + 1] = arr[j]
            j -= 1            
        arr[j + 1] = temp

    return arr


if __name__ == "__main__":
    arr = [5, 1, 8, 4, 9, 3, 6, 4, 8, 11, 15, 10, 7]
    print(InsertionSort(arr))

File: insertion_sort/insertion_sort.cpp
#include <iostream>
#include <vector>

using namespace std;


template <typename RandomIt>
void insertion_sort(RandomIt begin, RandomIt end)
{
	for (RandomIt i = begin+1; i != end; i++)
	{
		RandomIt j = i - 1;
		typename RandomIt::value_type temp = *i;
		
		while ((*j > temp) && (j != begin))
		{
			*(j+1) = *j;
			j--;
		}

		if ((j == begin) && (*j > temp))
		{
			*(j+1) = *j;
			*j = temp;
		}
		else
		{
			*(j+1) = temp;
		}
	}
}


int main()
{
	vector<int> a = { 5, 1, 4, 3, 9, 6, 2, 8, 7 };
	
	insertion_sort(a.begin(), a.end());
	for (const auto& i : a)
		cout << i << " ";

	return 0;
}
